src/Simulation/QCTraceSimulator/SortedQubitPool.cs

```diff
@@ -0,0 +1,329 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+
+namespace Microsoft.Quantum.Simulation.QCTraceSimulatorRuntime {
+
+    /// <summary>
+    /// Either time interval or point in time on the timeline of the program.
+    /// </summary>
+    internal struct ComplexTime
+    {
+        /// <summary>
+        /// Tracer configuration interface uses type "double" to provide duration of gates (aka depth).
+        /// When a gate is performed, time advances by the depth of the gate. DepthTime is used to count
+        /// time cosistent with the depth configuration settings. As a result, if a gate time is set to zero,
+        /// DepthTime won't increase after such gate is performed.
+        /// </summary>
+        internal double DepthTime { get; private set; }
+
+        /// <summary>
+        /// TrailingZeroDepthGateCount counts number of zero-depth gates after last non-zero depth gate is performed.
+        /// Generally, gates that happen later on a qubit cannot be scheduled before gates on the same qubit.
+        /// This is tracked transitively via qubit availability time, but zero-depth gates do not allow such calculation.
+        /// The situation is resolved by having this time component - TrailingZeroDepthGateCount.
+        /// </summary>
+        private long TrailingZeroDepthGateCount;
+
+        internal ComplexTime(double time) {
+            DepthTime = time;
+            TrailingZeroDepthGateCount = 0;
+        }
+
+        private ComplexTime(double depthTime, long trailingZeroDepthGateCount) {
+            DepthTime = depthTime;
+            TrailingZeroDepthGateCount = trailingZeroDepthGateCount;
+        }
+
+        public override string ToString() {
+            return $"{DepthTime}d+{TrailingZeroDepthGateCount}";
+        }
+        /// <summary>
+        /// Add gate time to availability time. If gate time is zero, advance dependency time.
+        /// </summary>
+        /// <param name="gateTime">Gate time to advance by. Assumed to be precise for comparison with 0.</param>
+        /// <returns>ComplexTime advanced by provided gate time</returns>
+        internal ComplexTime AdvanceBy(double gateTime) {
+            if (gateTime == 0.0) {
+                return new ComplexTime(DepthTime, TrailingZeroDepthGateCount + 1);
+            }
+            return new ComplexTime(DepthTime + gateTime);
+        }
+
+        /// <summary>
+        /// Subtracts argument ComplexTime from "this" ComplexTime. Returns the result.
+        /// Depth times are assumed to be precise for comparison.
```

Comment:
Is it safe to assume so? If above we are likely dealing with constants here the time is result of double-precision math. Maybe, compare with tolerance or assert that when non-equal, the difference is above the tolerance threshold?

Reply:
This is one of the two concerns I have about this approach. User can set gate times to any floating number. I just assume that this number would be reasonable. If numbers are reasonable, such as integers, or even 1/2, or 7/8, this will work because floating point manipulations are only additions and subtractions. This may break with gate duration set to 1/3. The only real fix for this is to change gate time from floating point to integral part in the interface. I'm not sure checking against threshold is a good approach - there's not reasonable threshold here.